### PR TITLE
Tolerate empty reads in `quicvarint.Read`

### PR DESCRIPTION
**Pull request by bemasc**
Currently, `quicvarint.Read` misinterprets an empty read (`n==0`) as a single zero byte.  While empty reads are "discouraged" in the `io.Reader` interface, they do sometimes happen, especially when using `io.Pipe()`.

This change tolerates empty reads, adds a corresponding unit test, and also includes some additional test coverage related to empty capsules

### CODE REVIEW

File: quicvarint/io.go [hunk 1]
```diff
@@ -31,7 +31,12 @@ func NewReader(r io.Reader) Reader {
 
 func (r *byteReader) ReadByte() (byte, error) {
 	var b [1]byte
-	n, err := r.Read(b[:])
+	var n int
+	var err error
+	for n == 0 && err == nil {
+		n, err = r.Read(b[:])
```
[the remark this answers is not in the document]
Comment (marten-seemann) [reply quicvarint/io.go:37]: What a useless comment from the word salad machine...

**Comment by marten-seemann** [comment]
@bemasc Thank you for your PR! Does the reader only return `0, nil` once, or does it do that repeatedly? In that case, this PR will lead to busy-looping. This seems suboptimal, but might still be better than returning an incorrect result.

I extended the varint benchmark in #5278. Unfortunately, this PR slows down varint parsing:
```
❯ benchstat old.txt new.txt
name                        old time/op  new time/op  delta
ReadBytesReader/1-byte-16   3.95ns ± 3%  3.85ns ± 3%  -2.32%  (p=0.000 n=48+47)
ReadBytesReader/2-byte-16   5.74ns ± 3%  5.40ns ± 2%  -5.95%  (p=0.000 n=49+50)
ReadBytesReader/4-byte-16   8.13ns ± 1%  7.91ns ± 2%  -2.76%  (p=0.000 n=47+48)
ReadBytesReader/8-byte-16   13.1ns ± 3%  12.9ns ± 3%  -1.46%  (p=0.000 n=50+47)
ReadSimpleReader/1-byte-16  13.4ns ± 4%  13.9ns ± 2%  +3.54%  (p=0.000 n=50+50)
ReadSimpleReader/2-byte-16  23.8ns ± 3%  25.1ns ± 3%  +5.59%  (p=0.000 n=50+49)
ReadSimpleReader/4-byte-16  44.2ns ± 4%  46.7ns ± 2%  +5.43%  (p=0.000 n=50+49)
ReadSimpleReader/8-byte-16  84.8ns ± 4%  91.1ns ± 2%  +7.40%  (p=0.000 n=50+50)
```

Not really sure why `ReadBytesReader` is getting faster, there shouldn't be any change in this code path. Looks like the loop is costing us ~5%. I tried replacing it with a `goto`, but this resulted in the same slowdown.

I don't know if there's anything we can do about it.

**Review by marten-seemann — APPROVED** [review]

**Comment by bemasc** [comment]
> Does the reader only return 0, nil once, or does it do that repeatedly? 

In the case I encountered, `0, nil` returns are sporadic.  Returning it constantly would be an especially terrible behavior.

If performance for non-ByteReader inputs is a priority, we could create a specialized `quicvarint.Read()` for `io.Reader` instead of wrapping it into a `ByteReader`.
